cf-keycrypt/cf-keycrypt.c

```diff
@@ -0,0 +1,504 @@
+    {
+        for(res = result; res != NULL; res = res->ai_next)
+        {
+            inet_ntop(
+                res->ai_family,
+                get_in_addr((struct sockaddr *)res->ai_addr),
+                ipaddress,
+                sizeof(ipaddress));
+            snprintf(
+                buffer, BUFSIZE * sizeof(char), "%s/ppkeys/root-%s.pub",
+                WORKDIR, ipaddress);
+            if (file_exist(buffer))
+            {
+                freeaddrinfo(result);
+                return buffer;
+            }
+        }
+    }
+    return NULL;
+}
+
+RSA *readseckey(const char *privkey_path)
+{
+    FILE *fp = fopen(privkey_path,"r");
+
+    if (fp == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not open private key '%s'", privkey_path);
+        return NULL;
+    }
+    RSA *privkey = PEM_read_RSAPrivateKey(fp, (RSA **)NULL, NULL,
+                                          (void *) passphrase);
+    if (privkey == NULL)
+    {
+        unsigned long err = ERR_get_error();
+        Log(LOG_LEVEL_ERR, "Could not read private key '%s': %s",
+            privkey_path, ERR_reason_error_string(err));
+    }
+    fclose(fp);
+    return privkey;
+}
+
+RSA *readpubkey(const char *pubkey_path)
+{
+    FILE *fp = fopen(pubkey_path, "r");
+
+    if (fp == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not open public key '%s'", pubkey_path);
+        return NULL;
+    }
+
+    RSA *pubkey = PEM_read_RSAPublicKey(
+        fp, NULL, NULL, (void *)passphrase);
+    if (pubkey == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not read public key '%s'", pubkey_path);
+    }
+    fclose(fp);
+    return pubkey;
+}
+
+long int rsa_encrypt(
+    const char *pubkey_path, const char *input_path, const char *output_path)
+{
+    int ks = 0;
+    unsigned long len = 0, ciphsz = 0;
+
+    RSA* pubkey = readpubkey(pubkey_path);
+    if (pubkey == NULL)
+    {
+        return -1;
+    }
+
+    FILE *input_file = fopen(input_path, "r");
+    if (input_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not open input file '%s'", input_path);
+        RSA_free(pubkey);
+        return -1;
+    }
+
+    FILE *output_file = fopen(output_path, "w");
+    if (output_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not create output file '%s'", output_path);
+        fclose(input_file);
+        RSA_free(pubkey);
+        return -1;
+    }
+
+    ks = RSA_size(pubkey);
+    char tmp_ciphertext[ks], tmp_plaintext[ks];
+
+    srand(time(NULL));
+    bool error_return = false;
+    while (!feof(input_file))
+    {
+        memset(tmp_plaintext, '\0', ks);
+        memset(tmp_ciphertext, '\0', ks);
+        len = fread(
+            tmp_plaintext, 1, ks - RSA_PKCS1_PADDING_SIZE, input_file);
+        if (len <= 0 || ferror(input_file) != 0)
+        {
+            Log(LOG_LEVEL_ERR, "Could not read file '%s'", input_path);
+            error_return = true;
+            break;
+        }
+        unsigned long size = RSA_public_encrypt(
```

Comment:
`size` should be `int`, because that's what `RSA_public_encrypt` returns and because it's compared to negative numbers later on.